### PR TITLE
docs: flesh out how to handle resources on deletion.

### PR DESCRIPTION
**Pull request by fgiloux**
**Description of the change:**

This change aims to enhance the documentation by providing more detailed instructions under advanced topics on how to handle resource deletion, going through multiple scenarios.
It also add a couple of lines to the best practices section to refer to it.

**Motivation for the change:**

Handling resources on deletion is an important topic. Accumulating resources without boundaries can have a very negative impact not only for the culprit operator but the complete cluster.

### CODE REVIEW

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 1]
```diff
@@ -206,6 +216,15 @@ func (r *MemcachedReconciler) finalizeMemcached(reqLogger logr.Logger, m *cachev
 }
 ```
 
+**Complex cleanup logic**
+
+Similar to the previous scenario finalizers can be used for implementing complex cleanup logic. Taking cron jobs as an example. They require to keep a number of past jobs, which have been spawned by the cron job controller. Therefore the cron job resource defines two fields successfulJobsHistoryLimit and failedJobsHistoryLimit. This is the role of the cron job controller to check the number of successful and failed jobs for a specific cron job and to delete the older ones when the limit has been exceeded. Here is [an example from Kubebuilder](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#3-clean-up-old-jobs-according-to-the-history-limit) for such a logic.
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:221]:
```suggestion
Similar to the previous scenario, finalizers can be used for implementing complex cleanup logic. Take [CronJobs](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) as an example: the controller maintains limited-size lists of jobs that have been created by the CronJob controller to check for deletion. These list sizes are configured by the CronJob fields [`.spec.successfulJobsHistoryLimit` and `.spec.failedJobsHistoryLimit`](https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#jobs-history-limits), which specify how many completed and failed jobs should be kept. Check out the [Kubebuilder CronJob tutorial](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#3-clean-up-old-jobs-according-to-the-history-limit) for full implementation details.
```

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 2]
```diff
@@ -206,6 +216,15 @@ func (r *MemcachedReconciler) finalizeMemcached(reqLogger logr.Logger, m *cachev
...
+
+**Sensitive resources**
+
+Sensitive resources need to be protected against unintended deletion. One typical example is Persistent Volume (PV) / Persistent Volume Claim (PVC). If a user tries to delete a PV that is bound by a PVC the PV is not removed immediately. PV removal is postponed until the PV is not bound to any PVC. To achieve a similar behaviour for your own resources Finalizers again can be leveraged. By setting a Finalizer on the PV like resource you controller can make sure there is no remaining resource depending on it before completing the deletion.
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:225]:
```suggestion
Sensitive resources need to be protected against unintended deletion. An intuitive example of protecting resources is the [PersistentVolume (PV) / PersistentVolumeClaim (PVC)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) relationship. A PV is first created, after which users can request access to that PV's storage by creating a PVC, which gets bound to the PV. If a user tries to delete a PV currently bound by a PVC, the PV is not removed immediately. Instead, PV removal is postponed until the PV is not bound to any PVC. Finalizers again can be leveraged to achieve a similar behaviour for your own PV-like custom resources: by setting a finalizer on an object, your controller can make sure there are no remaining objects bound to it before removing the finalizer and deleting the object.
```

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 3]
```diff
@@ -206,6 +216,15 @@ func (r *MemcachedReconciler) finalizeMemcached(reqLogger logr.Logger, m *cachev
...
+In the other direction the user can specify what happens to a PV when the PVC is deleted through the reclaim policy. There are two options: "delete", which means that the PV is deleted or retain, where the PV and its associated storage are kept after the PVC deletion. Such a behaviour is achieved again through the use of Finalizers. Key is that the user can make a conscious decision between automated cleansing of unused resources (delete) and data protection (retain).
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:226]:
```suggestion
Additionally, the user who created the PVC can specify what happens to the underlying storage allocated in a PV when the PVC is deleted through the [reclaim policy](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaiming). There are several options available, each of which defines a behavior that is achieved again through the use of finalizers. The key concept to take away is that your operator can give a user the power to decide how their resources are cleaned up via finalizers, which may be dangerous yet useful depending on your workloads.
```

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 4]
```diff
@@ -120,14 +120,24 @@ To learn about how metrics work in the Operator SDK read the [metrics section][m
 
 ### Handle Cleanup on Deletion
 
-To implement complex deletion logic, you can add a finalizer to your Custom Resource. This will prevent your Custom Resource from being
-deleted until you remove the finalizer (ie, after your cleanup logic has successfully run). For more information, see the
-[official Kubernetes documentation on finalizers](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers).
+Operators may create objects as part of their operational duty. Object accumulation can consume unnecessary resources, slow down the API and clutter the user interface. As such it is important for operators to keep good hygiene and to clean up resources when they are not needed. Here are a few common scenarios.
 
-**Example:**
+**Internal Resources**
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:125]:
```suggestion
#### Internal Resources
```

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 5]
```diff
@@ -120,14 +120,24 @@ To learn about how metrics work in the Operator SDK read the [metrics section][m
...
+
+A typical example is jobs. When they get created one or multiple pods get spawn. When the jobs are deleted the associated pods are deleted as well. This is a very common pattern and it can easily be achieved by setting the owner reference on the child resource, the pod in this example. Here is a code snippet for it where "r" is the reconcilier and "ctrl" the controller-runtime library:
+
+```go
+ctrl.SetControllerReference(job, pod, r.Scheme)
+```
+
+Note that with CRD the default behavior for cascading deletion is background. [This page](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/) in kubernetes documentation provides additional information.
+
+**External Resources**
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:135]:
```suggestion
#### External Resources
```

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 6]
```diff
@@ -206,6 +216,15 @@ func (r *MemcachedReconciler) finalizeMemcached(reqLogger logr.Logger, m *cachev
 }
 ```
 
+**Complex cleanup logic**
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:219]:
```suggestion
#### Complex cleanup logic
```

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 7]
```diff
@@ -206,6 +216,15 @@ func (r *MemcachedReconciler) finalizeMemcached(reqLogger logr.Logger, m *cachev
...
+
+Similar to the previous scenario finalizers can be used for implementing complex cleanup logic. Taking cron jobs as an example. They require to keep a number of past jobs, which have been spawned by the cron job controller. Therefore the cron job resource defines two fields successfulJobsHistoryLimit and failedJobsHistoryLimit. This is the role of the cron job controller to check the number of successful and failed jobs for a specific cron job and to delete the older ones when the limit has been exceeded. Here is [an example from Kubebuilder](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#3-clean-up-old-jobs-according-to-the-history-limit) for such a logic.
+
+**Sensitive resources**
```
Comment (estroz) [inline website/content/en/docs/building-operators/golang/advanced-topics.md:223]:
```suggestion
#### Sensitive resources
```

**Review by estroz — CHANGES_REQUESTED** [review]
Great content, needs grammatical work though. Thanks for the contribution!

**Comment by fgiloux** [comment]
Thanks @estroz for correcting the grammar and improving readability. I have accepted all the recommendations.

**Review by estroz — APPROVED** [review]
/lgtm

File: website/content/en/docs/building-operators/golang/advanced-topics.md [hunk 8]
```diff
@@ -120,14 +120,24 @@ To learn about how metrics work in the Operator SDK read the [metrics section][m
...
+
+Sometimes external resources or resources that are not owned by a custom resource need to be cleaned up when the parent resource is deleted. In that case [Finalizers](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers) can be leveraged. The deletion of the object becomes an update where a deletion timestamp is set. Without finalizers the object would just be dropped of the cache. The reconciliation loop of the controller will then need to check whether a date has been set for the deleteion timestamp, perform the cleanup and remove the finalizers to allow the garbage collection of the object.
+
+The following is a snippet from a theoretical controller file `controllers/memcached_controller.go` that implements a finalizer handler:
 
-The following is a snippet from a theoretical controller file `controllers/memcached_controller.go`
-that implements a finalizer handler:
 
 ```go
 import (
```
[the remark this answers is not in the document]
Comment (rashmigottipati) [reply website/content/en/docs/building-operators/golang/advanced-topics.md:143]: @estroz sure, will create a PR to rewrite the example to use the finalizer library.

**Pull request merged**